modules/services/hercules-ci-agent/default.nix

```diff
@@ -0,0 +1,104 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.hercules-ci-agent;
+  user = config.users.users.hercules-ci-agent;
+in {
+  options.services.hercules-ci-agent = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Run Hercules CI Agent as a system service";
+    };
+
+    apiBaseUrl = mkOption {
+      description = "Alternative base URL for the Hercules API";
+      example = "https://hercules-ci.com";
+      default = null;
+      type = types.nullOr types.string;
+    };
+
+    clusterJoinTokenPath = mkOption {
+      description = ''
+        Location of a the cluster join token. It authorizes the agent to add
+        itself to the cluster that the token represents.
+
+        This file is only required to be present for the agent's first run. It
+        will be ignored after the agent has used the token successfully.
+      '';
+      type = types.path;
+    };
+
+    logFile = mkOption {
+      type = types.path;
+      default = "/var/log/hercules-ci-agent.log";
+      description = "Stdout and sterr of hercules-ci-agent process.";
+    };
+
+    concurrentTasks = mkOption {
+      description = "Number of tasks to perform simultaneously, such as evaluations, derivations";
+      type = types.int;
+      default = 4;
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = (import (builtins.fetchTarball "https://github.com/hercules-ci/hercules-ci-agent/archive/stable.tar.gz")).nixos-19_03.hercules-ci-agent;
+      description = ''
+        Package containing the bin/hercules-ci-agent program.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    launchd.daemons.hercules-ci-agent = {
+      script = ''
+        exec ${cfg.package}/bin/hercules-ci-agent \
+          ${if (cfg.apiBaseUrl == null) then "" else "--api-base-url ${escapeShellArg cfg.apiBaseUrl}"} \
+          --cluster-join-token-path ${escapeShellArg cfg.clusterJoinTokenPath} \
+          --concurrent-tasks ${toString cfg.concurrentTasks}
+      '';
+
+      path = [ config.nix.package ];
+      environment = {
+        SSL_CERT_FILE = "${pkgs.cacert}/etc/ssl/certs/ca-bundle.crt";
+      };
+
+      serviceConfig.KeepAlive = true;
+      serviceConfig.RunAtLoad = true;
+      serviceConfig.StandardErrorPath = cfg.logFile;
+      serviceConfig.StandardOutPath = cfg.logFile;
+      serviceConfig.GroupName = "hercules-ci-agent";
+      serviceConfig.UserName = "hercules-ci-agent";
+      serviceConfig.WorkingDirectory = user.home;
+      serviceConfig.WatchPaths = [
+        cfg.clusterJoinTokenPath
+      ];
+    };
+
+    system.activationScripts.preActivation.text = ''
+      touch '${cfg.logFile}'
+      chown ${toString user.uid}:${toString user.gid} '${cfg.logFile}'
+    '';
+
+    users.knownGroups = [ "hercules-ci-agent" ];
+    users.knownUsers = [ "hercules-ci-agent" ];
```

Comment:
These options are intended for the user config, the idea was to have the user specify a whitelist of users that are safe to manage. The ofborg module asserts these are set instead.

Reply:
I'm not entirely sure I understand the motivation - safe from what point of view?

Agent user should always be automatically managed, or is just adding users in general an insecure setting?

Maybe the activation scripts could print what users were added (IIRC they already do that).

Reply:
I definitively want to get rid if this at some point, but the users module is currently pretty naive.  If it's not specified outside of the module the user would never get deleted.